src/preprocess/preprocess_node.cpp

```diff
@@ -1,51 +1,50 @@
+#include <ros/ros.h>
 #include <cv_bridge/cv_bridge.h>
 #include <image_transport/image_transport.h>
-#include <ros/ros.h>
 #include <sensor_msgs/image_encodings.h>
-#include <opencv2/core/core.hpp>
-#include <opencv2/highgui/highgui.hpp>
-#include <opencv2/imgcodecs.hpp>
+#include <sensor_msgs/Image.h>
 
 #include "median_subtract.h"
 #include "whitebalance.h"
 
-using namespace cv;
+using cv::Mat;
+
+image_transport::Publisher pub;
+image_transport::Subscriber sub;
+
+
+void process(Mat &img) {
+  // img is in BGR
+  // Do any processing and place the result in img
+
+  cvtColor(img, img, CV_BGR2HSV);
+  whiteBalance(img);
+  subtractMedian(img);
+  cvtColor(img, img, CV_HSV2BGR);
+}
+
+void callback(const sensor_msgs::ImageConstPtr &msg) {
+  cv_bridge::CvImagePtr cv_ptr;
+  try {
+    cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+  } catch (cv_bridge::Exception &e) {
+    ROS_ERROR("cv_bridge exception: %s", e.what());
+    return;
+  }
+  
+  process(cv_ptr->image);
+  pub.publish(cv_ptr->toImageMsg());
+}
+
 int main(int argc, char** argv) {
   ros::init(argc, argv, "image_preprocessor");
-  std::string imageName("");
-  if (argc > 1) {
-    imageName = argv[1];
-  }
 
-  Mat image;
-  image = imread(imageName.c_str(), IMREAD_COLOR);
+  ros::NodeHandle nh;
+  image_transport::ImageTransport it(nh);
 
-  if (image.empty()) {
-    std::cout << "Could not open or find the image" << std::endl;
-    return -1;
-  }
+  sub = it.subscribe("input", 1, callback);
```

Comment:
read topics from launch files

Reply:
I'm having the launch file do remapping instead. I think it's less error prone and more standard

Reply:
oh ok. that also works.